### ISSUE TITLE
Experiments running the build in parallel - Parallel tests fail

### ISSUE THREAD

**Issue by Leward** [spring-projects/spring-petclinic#277]
Hello, 

I am trying to play around with maven to experiment where I could parallelize the builds in order to take advantage of the latest hardware (fast SSD, a lot of cores, etc.).

For that, I picked the Spring Pet Clinic to play around and I got issues when I try to run the tests in parallel through the maven-surefire-plugin.

This is repository and the branch I push the code: https://github.com/Leward/spring-petclinic/tree/parallel

`mvn clean package` will trigger the error (given you have more than 1 core/thread available)

The test failture I encountered so far are due to: 

```
017-10-30 22:42:52.797  WARN 20507 --- [pool-2-thread-4] o.s.test.context.TestContextManager      : Caught exception while allowing TestExecutionListener [org.springframework.boot.test.autoconfigure.web.servlet.MockMvcPrintOnlyOnFailureTestExecutionListener@13d9b21f] to process 'after' execution for test: method [public void org.springframework.samples.petclinic.vet.VetController7Tests.testShowVetListXml() throws java.lang.Exception], instance [org.springframework.samples.petclinic.vet.VetController7Tests@3ac4f0cc], exception [java.lang.AssertionError: Body content
Expected: an XML document with XPath /vets/vetList[id=1]/id 
     but: xpath returned no results.]

java.util.ConcurrentModificationException: null
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:901) ~[na:1.8.0_131]
	at java.util.ArrayList$Itr.next(ArrayList.java:851) ~[na:1.8.0_131]
	at org.springframework.boot.test.autoconfigure.web.servlet.SpringBootMockMvcBuilderCustomizer$SystemLinesWriter.write(SpringBootMockMvcBuilderCustomizer.java:307) ~[spring-boot-test-autoconfigure-1.5.4.RELEASE.jar:1.5.4.RELEASE]
	at org.springframework.boot.test.autoconfigure.web.servlet.SpringBootMockMvcBuilderCustomizer$DeferredLinesWriter.writeDeferredResult(SpringBootMockMvcBuilderCustomizer.java:257) ~[spring-boot-test-autoconfigure-1.5.4.RELEASE.jar:1.5.4.RELEASE]
	at org.springframework.boot.test.autoconfigure.web.servlet.MockMvcPrintOnlyOnFailureTestExecutionListener.afterTestMethod(MockMvcPrintOnlyOnFailureTestExecutionListener.java:38) ~[spring-boot-test-autoconfigure-1.5.4.RELEASE.jar:1.5.4.RELEASE]
	at org.springframework.test.context.TestContextManager.afterTestMethod(TestContextManager.java:319) ~[spring-test-4.3.9.RELEASE.jar:4.3.9.RELEASE]
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:94) [spring-test-4.3.9.RELEASE.jar:4.3.9.RELEASE]
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84) [spring-test-4.3.9.RELEASE.jar:4.3.9.RELEASE]
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325) [junit-4.12.jar:4.12]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:252) [spring-test-4.3.9.RELEASE.jar:4.3.9.RELEASE]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:94) [spring-test-4.3.9.RELEASE.jar:4.3.9.RELEASE]
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) [junit-4.12.jar:4.12]
	at org.apache.maven.surefire.junitcore.pc.Scheduler$1.run(Scheduler.java:387) [surefire-junit47-2.18.1.jar:2.18.1]
	at org.apache.maven.surefire.junitcore.pc.InvokerStrategy.schedule(InvokerStrategy.java:54) [surefire-junit47-2.18.1.jar:2.18.1]
	at org.apache.maven.surefire.junitcore.pc.Scheduler.schedule(Scheduler.java:346) [surefire-junit47-2.18.1.jar:2.18.1]
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) [junit-4.12.jar:4.12]
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61) [spring-test-4.3.9.RELEASE.jar:4.3.9.RELEASE]
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70) [spring-test-4.3.9.RELEASE.jar:4.3.9.RELEASE]
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363) [junit-4.12.jar:4.12]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:191) [spring-test-4.3.9.RELEASE.jar:4.3.9.RELEASE]
	at org.junit.runners.Suite.runChild(Suite.java:128) [junit-4.12.jar:4.12]
	at org.junit.runners.Suite.runChild(Suite.java:27) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) [junit-4.12.jar:4.12]
	at org.apache.maven.surefire.junitcore.pc.Scheduler$1.run(Scheduler.java:387) [surefire-junit47-2.18.1.jar:2.18.1]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_131]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_131]

Tests run: 4, Failures: 1, Errors: 1, Skipped: 0, Time elapsed: 10.476 sec <<< FAILURE! - in org.springframework.samples.petclinic.vet.VetController7Tests
testShowVetListXml(org.springframework.samples.petclinic.vet.VetController7Tests)  Time elapsed: 5.075 sec  <<< FAILURE!
```

or 

```
Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 9.679 sec <<< FAILURE! - in org.springframework.samples.petclinic.vet.VetController7Tests
testShowResourcesVetList(org.springframework.samples.petclinic.vet.VetController7Tests)  Time elapsed: 9.529 sec  <<< FAILURE!
java.lang.AssertionError: No value at JSON path "$.vetList[0].id", exception: No results for path: $['vetList'][0]['id']
	at org.springframework.test.util.JsonPathExpectationsHelper.evaluateJsonPath(JsonPathExpectationsHelper.java:245)
	at org.springframework.test.util.JsonPathExpectationsHelper.assertValue(JsonPathExpectationsHelper.java:99)
	at org.springframework.test.web.servlet.result.JsonPathResultMatchers$2.match(JsonPathResultMatchers.java:100)
	at org.springframework.test.web.servlet.MockMvc$1.andExpect(MockMvc.java:171)
	at org.springframework.samples.petclinic.vet.VetController7Tests.testShowResourcesVetList(VetController7Tests.java:63)
```

Do you have any idea? It always fails on the VetControllerX test. (I copied it several times to emulate the fact I have a bunch of tests to run).

Could it be there is an issue with mockMvc not being thread safe?

**Comment by dsyer** [comment]
It's certainly not an issue with the petclinic code. Maybe open a new issue in Spring Boot?

**Comment by vfedoriv** [comment]
You have one mock bean for all tests.  And this bean not tread-safe.
